java/ql/src/experimental/semmle/code/java/security/StaticInitializationVectorQuery.qll

```diff
@@ -0,0 +1,166 @@
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.TaintTracking2
+
+/**
+ * Holds if `array` is initialized only with constants, for example,
+ * `new byte[8]` or `new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 }`.
+ */
+private predicate initializedWithConstants(ArrayCreationExpr array) {
+  not exists(array.getInit())
+  or
+  forex(Expr element | element = array.getInit().getAChildExpr() |
+    element instanceof CompileTimeConstantExpr
+  )
+}
+
+/**
+ * An expression that creates a byte array that is initialized with constants.
+ */
+private class StaticByteArrayCreation extends ArrayCreationExpr {
+  StaticByteArrayCreation() {
+    this.getType().(Array).getElementType().(PrimitiveType).getName() = "byte" and
+    initializedWithConstants(this)
+  }
+}
+
+/** Defines a sub-set of expressions that update an array. */
+private class ArrayUpdate extends Expr {
+  Expr array;
+
+  ArrayUpdate() {
+    exists(Assignment assign, ArrayAccess arrayAccess | arrayAccess = assign.getDest() |
+      assign = this and
+      arrayAccess.getArray() = array and
+      not assign.getSource() instanceof CompileTimeConstantExpr
+    )
+    or
+    exists(StaticMethodAccess ma |
+      ma.getMethod().hasQualifiedName("java.lang", "System", "arraycopy") and
+      ma = this and
+      ma.getArgument(2) = array
+    )
+    or
+    exists(StaticMethodAccess ma |
+      ma.getMethod().hasQualifiedName("java.util", "Arrays", "copyOf") and
```

Comment:
This is not really updating an existing array, instead it creates a copy of it. Might be good to add a comment here explaining the rationale since otherwise the class name `ArrayUpdate` is rather misleading.

If you want to cover copying an existing array then this might also have to consider `Arrays.copyOfRange` and calls to `clone()` of an array type.

Reply:
Also interested in the answer to this

Reply:
> This is not really updating an existing array, instead it creates a copy of it. Might be good to add a comment here explaining the rationale since otherwise the class name `ArrayUpdate` is rather misleading.

The class `ArrayUpdate` covers two cases:

1. when content of the array is updated, for example, by `System.arraycopy()`
2. when the array reference is updated, for example, by `Arrays.copyOf()`

I can update the qldoc for the class with the above. I can also split the class to two -- one class for each case.

I agree that `ArrayUpdate` may look misleading. Unfortunately, I didn't find a better name. Do you have a better name?

> If you want to cover copying an existing array then this might also have to consider `Arrays.copyOfRange` and calls to `clone()` of an array type.

Sure, I can cover these methods as well.

Reply:
I don't think you need to model any of these functions: they naturally interrupt dataflow from a static array to its usage by overwriting the variable or field the reference is stored in.

If a dataflow path is found anyhow then most likely there is a control-flow route around the assignment and it's genuinely possible for the static array to be used as an IV.

Reply:
> I don't think you need to model any of these functions: they naturally interrupt dataflow from a static array to its usage by overwriting the variable or field the reference is stored in.

If I remember correctly, while testing the query, these functions didn't interrupt dataflow. That sometimes resulted in some false-positives. But I might be missing something. I did a quick test and looks like you're right. I'll remove these functions.